Use `CONFIG_(BIG|LITTLE)_ENDIAN` instead of `__BYTE_ORDER__` for endianness checks

carlescufi opened the pull request:
See commit messages.

**Please Ignore the compliance check failure, it's just checkpatch doing its job with the new check**

Note that this PR includes the following fix: https://github.com/littlefs-project/littlefs/pull/742

carlescufi committed:
toolchain: Ensure that Kconfig endianness matches the compiler's

In preparation for using CONFIG_(BIG|LITTLE)_ENDIAN exclusively for
endianness-based conditional compilation in the tree, ensure that the
Kconfig CONFIG_(BIG|LITTLE)_ENDIAN options match the compiler's
__BYTE_ORDER__ macro.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
treewide: Use CONFIG_*_ENDIAN instead of __BYTE_ORDER__

In order to avoid using multiple sources of truth for the platfom's
endianness, convert the in-tree code to use the (BIG|LITTLE)_ENDIAN
Kconfig variables exclusively, instead of the compiler's
__BYTE_ORDER__.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

gmarull commented on drivers/w1/w1_net.c:
```diff
@@ -183,7 +183,7 @@ int z_impl_w1_search_bus(const struct device *dev, uint8_t command,
 		 * bytes are stored in big-endian byte order.
 		 */
-		if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {
+		if (CONFIG_BIG_ENDIAN) {
```
shouldn't this be using `IS_ENABLED`?

carlescufi replied:
ooops, thanks!

gmarull commented on modules/littlefs/zephyr_lfs_config.h:
```diff
@@ -147,21 +147,9 @@ static inline int lfs_scmp(uint32_t a, uint32_t b)
-	__BYTE_ORDER == __ORDER_BIG_ENDIAN) ||                       \
-	(defined(__BYTE_ORDER__) && defined(__ORDER_BIG_ENDIAN__) && \
-	__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__))
+#elif !defined(LFS_NO_INTRINSICS)
```
`&& defined(CONFIG_BIG_ENDIAN)`?

carlescufi replied:
Hmm no, because it´s an `#elif`?

gmarull replied:
Before:

```
if !LFS_NO_INTRINSICS && !CONFIG_BIG_ENDIAN
  return a;
elif !LFS_NO_INTRINSICS && CONFIG_BIG_ENDIAN
  return __builtin_bswap32(a);
else
  ...
```

Now:

```
if !CONFIG_BIG_ENDIAN
  return a;
elif !LFS_NO_INTRINSICS
  return __builtin_bswap32(a);
else:
 ...
```

If we now have: `LFS_NO_INTRINSICS && !CONFIG_BIG_ENDIAN`, we'll `return a`, whereas before the `else` branch would have been taken, and I think that is what is desired according to the description:

```c
// Builtin functions, these may be replaced by more efficient
// toolchain-specific implementations. LFS_NO_INTRINSICS falls back to a more
// expensive basic C implementation for debugging purposes
```

gmarull replied:
It looks like the previous behavior was actually wrong, so please add a first commit fixing the issue.

gmarull commented on include/zephyr/toolchain.h:
```diff
@@ -117,15 +117,22 @@
+ * and that they match the Kconfig option that is used in the code itself to
+ * check for endianness.
  */
 #ifndef _LINKER
 #if !defined(__BYTE_ORDER__) || !defined(__ORDER_BIG_ENDIAN__) || \
     !defined(__ORDER_LITTLE_ENDIAN__)
 
 #error "__BYTE_ORDER__ is not defined"
```
just a question here, is this check needed anymore?

carlescufi replied:
It's a double-check but I will defer to @tejlmand.

stephanosio replied:
yes, as long as we use `__BYTE_ORDER__` and its friends in tree (i.e. right below).

carlescufi replied:
Ah right, that's a good reason too :) but we could stop using them in the in-tree toolchain headers entirely if we wanted to (i.e. simplify even further), but I thought that was going too far.

gmarull replied:
If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right? So, the message above doesn't tell me anything the compiler can't.

stephanosio replied:
> If not defined, compiler will also complain about `__BYTE_ORDER__` not defined in the checks below, right?

Unless you specify `-Wundef`, no.

Moreover, `-Wundef` is not enabled by default nor is it enabled by `-Wall`.

stephanosio commented on drivers/w1/w1_net.c:
```diff
@@ -183,7 +183,7 @@ int z_impl_w1_search_bus(const struct device *dev, uint8_t command,
 		 * bytes are stored in big-endian byte order.
 		 */
-		if (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__) {
+		if (CONFIG_BIG_ENDIAN) {
```
I agree that attempting to support `__ORDER_PDP_ENDIAN__` is absurd. I have not seen a single architecture that uses `__ORDER_PDP_ENDIAN__` other than PDP-11, and I am pretty sure Zephyr is never going to run on PDP-11.

de-nordic replied:
> I agree that attempting to support `__ORDER_PDP_ENDIAN__` is absurd. I have not seen a single architecture that uses `__ORDER_PDP_ENDIAN__` other than PDP-11, and I am pretty sure Zephyr is never going to run on PDP-11.

You are breaking my heart. Still got some hopes for VAX-11...

de-nordic reviewed (dismissed):
mgmt/mcumgr is OK with the changes.

de-nordic commented:
Btw. failing CI needs this fix https://github.com/zephyrproject-rtos/zephyr/pull/51656

henrikbrixandersen commented:
Noting that we will likely still have lots of external code (modules) depending on the compiler definitions, what is the real gain here? Is this preparation for something else?

henrikbrixandersen commented:
Also, if we intend to only allow one truth for endianess in-tree we ought to add a checkpatch check or usage of the compiler definitions will slowly creep back in.

carlescufi commented:
> Noting that we will likely still have lots of external code (modules) depending on the compiler definitions, what is the real gain here? Is this preparation for something else?

The gain is to use Kconfig internally in-tree, because we have the Kconfig option. The fact that modules will continue to use `__BYTE_ORDER__` is not a reason not to be consistent ourselves, like we do in many other similar cases, eg. `__packed`.

carlescufi commented:
> Also, if we intend to only allow one truth for endianess in-tree we ought to add a checkpatch check or usage of the compiler definitions will slowly creep back in.

Yep, I thought about that and I agree. @stephanosio any suggestions here?

stephanosio commented:
> Yep, I thought about that and I agree. @stephanosio any suggestions here?

We can add a checkpatch check for it, something like:

https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563

carlescufi commented:
> > Yep, I thought about that and I agree. @stephanosio any suggestions here?
> 
> We can add a checkpatch check for it, something like:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563

Thanks! Will add this to the PR

cfriedt commented:
Just for clarification, we will still have `__BYTE_ORDER__`, `__ORDER_LITTLE_ENDIAN__`, and `__ORDER_BIG_ENDIAN__` defined by `toolchain.h`, correct?

It's just that I've spent the better part of a year, on and off, working on integrating Apache Thrift, and finally got this change upstream. It would be annoying if it were suddenly invalid.

https://github.com/apache/thrift/pull/2681/files

carlescufi committed:
checkpatch: Disallow using __BYTE_ORDER__

The canonical way of determining endianness is now CONFIG_BIG_ENDIAN
instead in Zephyr.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
> > > Yep, I thought about that and I agree. @stephanosio any suggestions here?
> > 
> > 
> > We can add a checkpatch check for it, something like:
> > https://github.com/zephyrproject-rtos/zephyr/blob/21c7e3ba65c20fa6cd574751780a010e8adb3dfa/scripts/checkpatch.pl#L6559-L6563
> 
> Thanks! Will add this to the PR

@henrikbrixandersen @stephanosio done, thanks!

Thalley reviewed (dismissed):
ACK for BLE Audio

rlubos reviewed (dismissed):
+1 for networking

henrikbrixandersen reviewed (changes requested):
I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to  `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

carlescufi commented:
> I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option. I think it'd be confusing for people to have the big endian version of a bitfield section in a struct first, because nowadays people instinctively recognize the first block to be the common/default case, which is little endian.

jhedberg commented:
> I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option

I think it's just general good practice, that if you have an else-branch, that if possible, the if-condition should be testing for the existence of something rather than the non-existence of something. At least that computes much faster in my brain :) It doesn't really matter what the real world frequency of occurrence of the "something" behind the condition is. Just my two cents.. :)

carlescufi commented:
> > I disagree. Most platforms are little endian, and thus for me it makes sense to have the little endian version at the top, as the "default" option
> 
> I think it's just general good practice, that if you have an else-branch, that if possible, the if-condition should be testing for the existence of something rather than the non-existence of something else. At least that computes much faster in my brain :) It doesn't really matter what the real world frequency of occurrence of the "something" behind the condition is. Just my two cents.. :)

I agree with that, but the question is whether we really want to have:

```
#ifdef CONFIG_BIG_ENDIAN
<big endian code that will almost never be used>
#else
<little endian code that will almost always be used>
#endif
```
I personally would not, but we are in the realm of personal preference here.
Perhaps it would help if, instead of `#ifndef`, I used `#if !defined()` ?

henrikbrixandersen commented:
> I agree with that, but the question is whether we really want to have:
> 
> ```
> #ifdef CONFIG_BIG_ENDIAN
> <big endian code that will almost never be used>
> #else
> <little endian code that will almost always be used>
> #endif
> ```

That reads much more naturally - at least my somewhat twisted mind :) The less negations, the easier it is for me to grasp.

> I personally would not, but we are in the realm of personal preference here. Perhaps it would help if, instead of `#ifndef`, I used `#if !defined()` ? @jhedberg and @henrikbrixandersen

I don't think using `#if !defined()` improves the readability, no.

henrikbrixandersen commented on modules/canopennode/CO_driver_target.h:
```diff
@@ -39,12 +39,10 @@ extern "C" {
+#ifndef CONFIG_BIG_ENDIAN
 #define CO_LITTLE_ENDIAN
-#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
-#define CO_BIG_ENDIAN
 #else
-#error "Unsupported endianness"
+#define CO_BIG_ENDIAN
 #endif
```
Take for instance this change. If `CONFIG_BIG_ENDIAN` is not defined, we define `CO_LITTLE_ENDIAN`. Otherwise we define `CO_BIG_ENDIAN`. This would read much naturally as:

```
#ifdef CONFIG_BIG_ENDIAN
#define CO_BIG_ENDIAN
#else
#define CO_LITTLE_ENDIAN
#endif
```

The same logic - in my mind - applies to all the other `#ifndef ... #else ... #endif` statements, but the above is a very literal example.

carlescufi replied:
Maybe we could go case by case then. Since this is mostly about personal preference, I can change this particular one to match what you describe, and leave the other as `#if !defined()` unless a particular maintainer asks me to change it.

rlubos commented:
> I personally would not, but we are in the realm of personal preference here.

Totally. While it doesn't really make much difference to me, I'd slightly lean towards @henrikbrixandersen proposal, just feels more natural to see big endian stuff after seeing CONFIG_BIG_ENDIAN, it reads easier.

carlescufi committed:
modules: littlefs: Fix the endianness conversion functions

See full description in the upstream PR:
https://github.com/littlefs-project/littlefs/pull/742

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

cvinayak commented:
> I would invert all of these `#ifndef CONFIG_BIG_ENDIAN ... #else ... #endif` sections to `#ifdef CONFIG_BIG_ENDIAN ... #else ... #endif` in order to make the code more readable. I think the current approach with mechanically replacing makes the code much less logical/readable than before.

It is naturally (easily) readable when `if (true)` is followed by the true statements. And looking for big-endian code below `CONFIG_BIG_ENDIAN` is much easier without training the brain to understand the opposite as have to carefully read `#ifndef` or `#if !defined`.

My suggestion is to add a hidden
```
config LITTLE_ENDIAN
        # Hidden Kconfig option representing the default little-endian architecture
        bool
        depends on !BIG_ENDIAN
        default y
```

and use:
```
#if defined(CONFIG_LITTLE_ENDIAN)
/* Little-endian code */
#else
/* Big-endian code */
#endif
```

carlescufi commented:
> My suggestion is to add a hidden
> 
> ```
> config LITTLE_ENDIAN
>         # Hidden Kconfig option representing the default little-endian architecture
>         bool
>         depends on !BIG_ENDIAN
>         default y
> ```

I thought about this, and so did @gmarull, but we thought that this was "cheating". @cvinayak is there a precedent for a hidden Kconfig like this that reflects the opposite of the visible one?

henrikbrixandersen commented:
> I thought about this, and so did @gmarull, but we thought that this was "cheating". @cvinayak @Thalley @gmarull @henrikbrixandersen @rlubos is there a precedent for a hidden Kconfig like this that reflects the opposite of the visible one?

That sounds more like a Kconfig `choice` to me.

stephanosio commented:
> My suggestion is to add a hidden
> 
> ```
> config LITTLE_ENDIAN
>         # Hidden Kconfig option representing the default little-endian architecture
>         bool
>         depends on !BIG_ENDIAN
>         default y
> ```
> 
> and use:

I agree that having both `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN` is nicer and makes more sense.

Regarding its implementation, we should be using `choice`, however (`choice` will ensure mutual exclusivity between `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN`).

carlescufi commented:
> Regarding its implementation, we should be using `choice`, however (`choice` will ensure mutual exclusivity between `CONFIG_LITTLE_ENDIAN` and `CONFIG_BIG_ENDIAN`).

Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?

stephanosio commented:
> Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?

No? The default will be `CONFIG_LITTLE_ENDIAN=y`. Setting `CONFIG_BIG_ENDIAN=y` will automatically set `CONFIG_LITTLE_ENDIAN=n` if implemented as a choice.

Something like:

```
choice
	default LITTLE_ENDIAN

config LITTLE_ENDIAN
	bool

config BIG_ENDIAN
	bool

endchoice
```

carlescufi commented:
> > Will that break backwards compatibility though then? i.e. if users were setting `CONFIG_BIG_ENDIAN=y` out-of-tree?
> 
> No? The default will be `CONFIG_LITTLE_ENDIAN=y`. Setting `CONFIG_BIG_ENDIAN=y` will automatically set `CONFIG_LITTLE_ENDIAN=n` if implemented as a choice.
> 
> Something like:
> 
> ```
> choice
> 	default LITTLE_ENDIAN
> 
> config LITTLE_ENDIAN
> 	bool
> 
> config BIG_ENDIAN
> 	bool
> 
> endchoice
> ```

Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

henrikbrixandersen commented:
> Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

Yes, but please use a named `choice` (perhaps `CONFIG_ENDIANNESS`?).

tejlmand commented:
> Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?

@carlescufi  please do not change this into a choice.
The current `BIG_ENDIAN` is promptless and thereby cannot be selected by the user, but is selected by the arch or the board, for example:
https://github.com/zephyrproject-rtos/zephyr/blob/fd5fe6ca8e8fbc89263ad16f78403989e1308870/arch/Kconfig#L67

But a choice must have a prompt, and thus can suddenly be changed by the user.

Furthermore, other symbols cannot do a select on a choice symbol.

stephanosio commented:
> But a choice must have a prompt, and thus can suddenly be changed by the user.

But that is not necessarily a bad thing? In fact, if you are working with a bi-endian architecture, users would want to be able to configure this.

> Furthermore, other symbols cannot do a select on a choice symbol.

This can easily be worked around by setting default for the choice in defconfig.

nordicjm commented:
> This can easily be worked around by setting default for the choice in defconfig.

As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot

carlescufi commented:
@tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?

nordicjm commented:
> Thanks @nordicjm! Do you have any objections against https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549?

That looks good!

carlescufi commented:
> > As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot
> 
> well, then that sounds like something that needs to be fixed for v2.
> 
> p.s. arch symbols should not be selecting endianness.

I don´t want to speak for @nordicjm and @tejlmand, but board v2 is fundamentally different (and opt-in) in the way it determines architecture, so it might be that the `.defconfig` behavior here is intentional

cfriedt commented:
> > Right, sorry, my confusion. @henrikbrixandersen @rlubos you OK with this?
> 
> Yes, but please use a named `choice` (perhaps `CONFIG_ENDIANNESS`?).

Yoda would say `CONFIG_ENDIAN_BIG` or `CONFIG_ENDIAN_LITTLE`... and then the choice could be `CONFIG_ENDIAN_CHOICE`.

![Yoda](https://media.giphy.com/media/8hMD9YakVza3452SpN/giphy.gif)

tejlmand commented:
> But that is not necessarily a bad thing? In fact, if you are working with a bi-endian architecture, users would want to be able to configure this.

@stephanosio if you're working on such an architecture the architecture itself should define a choice which can then select the proper endianess based on user selection.

No reason to expose such a choice to user on other architectures where this is not relevant.

tejlmand commented:
> > As part of boards v2, the defconfig fill will not be loaded at this point, so it cannot
> 
> well, then that sounds like something that needs to be fixed for v2.

I believe you're talking past each other.
The defconfig is part of Zephyr build Kconfig tree, not the SoC / arch trees.
So in a Zephyr build the defconfig is loaded, but in sysbuild (which can also use the arch / soc v2) there is no defconfig (as this file belongs to Zephyr).
Sysbuild has its own Kconfig tree, but shares part of it with Zephyr Kconfig tree.

But I guess further discussion of how this works is out of scope for this PR.

tejlmand commented:
> @tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?

yes, and it would even allow arch to do `select LITTLE_ENDIAN` although they won't have to.

henrikbrixandersen commented:
> > @tejlmand thanks for the input. So then @cvinayak's [proposal](https://github.com/zephyrproject-rtos/zephyr/pull/51666#issuecomment-1293278549) to add a hidden options instead could work, right?
> 
> 
> 
> yes, and it would even allow arch to do `select LITTLE_ENDIAN` although they won't have to.

Let's go with the two opposite options, then.

stephanosio commented:
> @stephanosio if you're working on such an architecture the architecture itself should define a choice which can then select the proper endianess based on user selection.

So your suggestion is to add `CONFIG_ARM_BIG_ENDIAN`, `CONFIG_ARM_LITTLE_ENDIAN`, `CONFIG_SPARC_BIG_ENDIAN`, `CONFIG_SPARC_LITTLE_ENDIAN`, `CONFIG_MIPS_BIG_ENDIAN`, `CONFIG_MIPS_LITTLE_ENDIAN` and so on? Because that sounds far from ideal to me.

> No reason to expose such a choice to user on other architectures where this is not relevant.

I guess it is sort of a double edged sword -- using `choice` gives you "unified configuration" for endianness across all architectures that can be user-overridden; at the same time, it opens up the possibility for users to override this setting for the architectures/SoCs that do not support bi-endianness.

Anyways, this use case seems to be beyond the scope of this PR. Let us revisit this when the need arises (but yeah, the `SPARC` symbol should really not be selecting `BIG_ENDIAN`, that is just plain wrong and we should fix it).

> So in a Zephyr build the defconfig is loaded, but in sysbuild (which can also use the arch / soc v2) there is no defconfig (as this file belongs to Zephyr).

That sounds like a problem to me; but, I do admit I have not given enough thought on this matter as you have; so, I assume this can be worked around somehow.

carlescufi committed:
arch: Add a CONFIG_LITTLE_ENDIAN hidden Kconfig option

In preparation for using CONFIG_*_ENDIAN instead of __BYTE_ORDER__, add
a reflection of CONFIG_BIG_ENDIAN that will allow users to conditionally
compile with #ifdef instead of #ifndef while keeping the default case
(little endian) as the first block in the conditional compilation.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@gmarull @cvinayak please take another look

gmarull commented on arch/Kconfig:
```diff
@@ -166,6 +166,14 @@ config BIG_ENDIAN
+	# conditional compilation
+        bool
+        depends on !BIG_ENDIAN
+        default y
```
mixes tabs/spaces

gmarull commented on modules/littlefs/zephyr_lfs_config.h:
```diff
@@ -178,26 +167,15 @@ static inline uint32_t lfs_tole32(uint32_t a)
-	       (((uint8_t *)&a)[1] << 16) |
-	       (((uint8_t *)&a)[2] <<  8) |
-	       (((uint8_t *)&a)[3] <<  0);
+	return (((uint8_t *)&a)[0] <<  0) |
+	       (((uint8_t *)&a)[1] <<  8) |
+	       (((uint8_t *)&a)[2] << 16) |
+	       (((uint8_t *)&a)[3] << 24);
```
hmm this is also being changed, was it wrong?

carlescufi replied:
how strange, no. This is an oversight. Thanks!

de-nordic reviewed (approved):
Mcumgr is OK.

Thalley reviewed (approved):
LE Audio ACK

cvinayak reviewed (approved):
`subsys/bluetooth/controller/ll_sw` approved.

henrikbrixandersen reviewed (approved):
Much more readable, thanks!

henrikbrixandersen commented on scripts/checkpatch.pl:
```diff
@@ -6562,6 +6562,12 @@ sub process {
+# check for uses of __BYTE_ORDER__
+		while ($line =~ /\b(__BYTE_ORDER__)\b/g) {
+			ERROR("BYTE_ORDER",
+			      "Use of the '$1' macro is disallowed. Use CONFIG_BIG_ENDIAN instead\n" . $herecurr);
```
```suggestion
			      "Use of the '$1' macro is disallowed. Use CONFIG_LITTLE_ENDIAN/CONFIG_BIG_ENDIAN instead\n" . $herecurr);
```

carlescufi replied:
Well spotted! will send a PR

henrikbrixandersen reviewed (approved):
Minor improvement.

marc-hb commented:
Ironically, commit ae123c9bd1ea1 ("toolchain: Ensure that Kconfig endianness matches the compiler's") triggers an endianness sparse error. Reverting ae123c9bd1ea1 makes the error go away.

```shell
$ west build -p -b intel_adsp_cavs25 samples/hello_world/ -- -DSPARSE=y | grep error:

/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"
/home/mherber2/SOF/zephyr/include/zephyr/toolchain.h:133:2: error: "Endiannes mismatch"

```

This could be a problem with sparse itself, I haven't investigated this _at all_, only  bisected it. Just a heads-up.

cc: @kv2019i , @juimonen , @lyakh , @lgirdwood
